README.pod
=pod

=encoding UTF-8

=head1 NAME

MooseX::Role::Parameterized - Moose roles with composition parameters

=head1 VERSION

version 1.11

=head1 SYNOPSIS

    package Counter;
    use MooseX::Role::Parameterized;

    parameter name => (
        isa      => 'Str',
        required => 1,
    );

    role {
        my $p = shift;

        my $name = $p->name;

        has $name => (
            is      => 'rw',
            isa     => 'Int',
            default => 0,
        );

        method "increment_$name" => sub {
            my $self = shift;
            $self->$name($self->$name + 1);
        };

        method "reset_$name" => sub {
            my $self = shift;
            $self->$name(0);
        };
    };

    package MyGame::Weapon;
    use Moose;

    with Counter => { name => 'enchantment' };

    package MyGame::Wand;
    use Moose;

    with Counter => { name => 'zapped' };

=head1 DESCRIPTION

Your parameterized role consists of two new things: parameter declarations
and a C<role> block.

Parameters are declared using the L</parameter> keyword which very much
resembles L<Moose/has>. You can use any option that L<Moose/has> accepts. The
default value for the C<is> option is C<ro> as that's a very common case. Use
C<< is => 'bare' >> if you want no accessor. These parameters will get their
values when the consuming class (or role) uses L<Moose/with>. A parameter
object will be constructed with these values, and passed to the C<role> block.

The C<role> block then uses the usual L<Moose::Role> keywords to build up a
role. You can shift off the parameter object to inspect what the consuming
class provided as parameters. You use the parameters to customize your
role however you wish.

There are many possible implementations for parameterized roles (hopefully with
a consistent enough API); I believe this to be the easiest and most flexible
design. Coincidentally, Pugs originally had an eerily similar design.

See L<MooseX::Role::Parameterized::Extending> for some tips on how to extend
this module.

=head2 Why a parameters object?

I've been asked several times "Why use a parameter I<object> and not just a
parameter I<hashref>? That would eliminate the need to explicitly declare your
parameters."

The benefits of using an object are similar to the benefits of using Moose. You
get an easy way to specify lazy defaults, type constraint, delegation, and so
on. You get to use MooseX modules.

=for Pod::Coverage current_metaclass method parameter role with

=head1 L<MooseX::Role::Parameterized::Tutorial>

B<Stop!> If you're new here, please read
L<MooseX::Role::Parameterized::Tutorial> for a much gentler introduction.

=for stopwords metaobject

You also get the usual introspective and intercessory abilities that come
standard with the metaobject protocol. Ambitious users should be able to add
traits to the parameters metaclass to further customize behavior. Please let
me know if you're doing anything viciously complicated with this extension. :)

=head1 CAVEATS

You must use this syntax to declare methods in the role block:
C<< method NAME => sub { ... }; >>. This is due to a limitation in Perl. In
return though you can use parameters I<in your methods>!

=head1 SEE ALSO

L<http://sartak.org/2009/01/parametric-roles-in-perl-5.html>

L<http://sartak.org/2009/05/the-design-of-parameterized-roles.html>

L<http://stevan-little.blogspot.com/2009/07/thoughts-on-parameterized-roles.html>

L<http://perldition.org/articles/Parameterized%20Roles%20with%20MooseX::Declare.pod>

L<http://www.modernperlbooks.com/mt/2011/01/the-parametric-role-of-my-mvc-plugin-system.html>

L<http://jjnapiorkowski.typepad.com/modern-perl/2010/08/parameterized-roles-and-method-traits-redo.html>

L<http://sartak.org/talks/yapc-asia-2009/(parameterized)-roles/>

=for stopwords Joose

L<https://github.com/SamuraiJack/JooseX-Role-Parameterized> - this extension ported to JavaScript's Joose

=head1 SUPPORT

Bugs may be submitted through L<the RT bug tracker|https://rt.cpan.org/Public/Dist/Display.html?Name=MooseX-Role-Parameterized>
(or L<bug-MooseX-Role-Parameterized@rt.cpan.org|mailto:bug-MooseX-Role-Parameterized@rt.cpan.org>).

There is also a mailing list available for users of this distribution, at
L<http://lists.perl.org/list/moose.html>.

There is also an irc channel available for users of this distribution, at
L<C<#moose> on C<irc.perl.org>|irc://irc.perl.org/#moose>.

=head1 AUTHOR

Shawn M Moore <code@sartak.org>

=head1 CONTRIBUTORS

=for stopwords Karen Etheridge Dave Rolsky Jesse Luehrs Oliver Charles Yuval Kogman Robert 'phaylon' Sedlacek Florian Ragwitz Mark Fowler Chris Weyl Csson Andy Jack Ricardo Signes Todd Hepler

=over 4

=item *

Karen Etheridge <ether@cpan.org>

=item *

Dave Rolsky <autarch@urth.org>

=item *

Jesse Luehrs <doy@tozt.net>

=item *

Oliver Charles <oliver.g.charles@googlemail.com>

=item *

Yuval Kogman <nothingmuch@woobling.org>

=item *

Robert 'phaylon' Sedlacek <rs@474.at>

=item *

Florian Ragwitz <rafl@debian.org>

=item *

Mark Fowler <mark@twoshortplanks.com>

=item *

Chris Weyl <cweyl@alumni.drew.edu>

=item *

Csson <erik.carlsson@live.com>

=item *

Andy Jack <github@veracity.ca>

=item *

Ricardo Signes <rjbs@cpan.org>

=item *

Todd Hepler <thepler@employees.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by Shawn M Moore.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Changes
Revision history for MooseX-Role-Parameterized

{{$NEXT}}

1.11      2019-07-16 01:33:16Z
   - remove MooseX::Role::WithOverloading from test dependencies (RT#130075)

1.10      2017-04-20 17:50:53Z
   - add MooseX::Role::Parameterised as stub documentation, for
     discoverability

1.09      2016-08-07 01:20:31Z
   - Give 'anonymous' generated roles names that are similar to the original
     parameterized role so debug output is easier to read

1.08      2014-08-23 22:38:12Z
    - add x_breaks metadata for incompatibility issue with MooseX::Storage
      (now resolved with MooseX-Storage-0.47)

1.07      2014-08-06 05:44:11Z
    - re-release to remove README.pod from shipped dist

1.06      2014-08-02 04:34:06Z
    - Restored MooseX::Role::Parameterized->current_metaclass as a public
      method. Apparently there is code on CPAN that relies on this.

1.05      2014-07-31 18:26:03Z
    - If a parameterizable role was reinitialized after any parameters or a
      role block was declared, those declarations were lost. Reinitialization
      usually occurs when new metaroles are applied to the role by other
      MooseX modules.

1.04      2014-07-31 14:53:26Z
    - Convert this distribution to Dist::Zilla to resolve packaging insanity

1.03  2014-07-30
    - This extension is now implemented as a role metarole, which means it can
      (mostly) cooperate with other role extensions, such as
      MooseX::Role::WithOverloading. Note that you should load
      MooseX::Role::Parameterized _after_ other extensions.
    - This module no longer supports passing a "-metaclass" parameter when you
      load it. This was an artifact from a much earlier era of Moose
      extensions.
    - repository migrated to the github moose organization

1.02  2013-09-11
    - set %INC entries in test to avoid failures with
      Moose pre-2.11 (kentfrederic)
        https://github.com/sartak/MooseX-Role-Parameterized/pull/7
        https://rt.cpan.org/Ticket/Display.html?id=88593

1.01  2013-09-10
    - removed use of deprecated Class::MOP::load_class (ether)

1.00  2012-01-12
    - MooseX::Role::Parameterized now depends on Moose 2.0300
      released 2011-09-23 for its several core improvements, hence the
      major version number bump.

    - MXRP now uses the new meta_lookup feature added to Moose 2.0300
      (specifically for MXRP) to avoid duplicating Moose-0.60-era
      Moose::Role sugar. This means the error messages you get are
      better, and there's a lot less ugly code in MXRP.
    - Remove alias/excludes special case errors. They have long since been
      replaced in core Moose by -alias and -excludes, and were finally
      removed in Moose 2.0200.
    - Expressly forbid using role { } inside role { } (which makes no
      sense but I suppose you could accidentally trigger it)
    - Documentation fix from cweyl@alumni.drew.edu
      https://github.com/sartak/MooseX-Role-Parameterized/pull/5

0.27  2011-08-22
    - Remove MooseX::Role::Parameterized::Meta::Parameter; instead,
      the default for parameters of (is => 'ro') is set by the
      "parameter" sugar. This removes an unnecessary metaclass and
      removes weird edge cases caused by it. Please please please
      report any breakage!

0.26  2011-03-20
    - make tests stop relying on a specific format for anon class/role names (doy)

0.25  2011-03-04
    - Move the reinitialization hook from Role to Trait so hardcore hackers
      (trait janitors) get the benefit too (doy)

0.24  2011-03-01
    - Made MXRP more extensible by hooking into the role metaobject
      reinitialization (doy)

0.23  2010-12-25
    - Merry Christmas!
    - Make MooseX::Role::Parameterized::Extending more instructive. Hopefully.

0.22  2010-11-26
    - The test suite now uses Test::Fatal instead of Test::Exception (Karen
      Etheridge).
    - Fix Test::More dependency (reported by Father Chrysostomos) [rt.cpan.org #63222]

0.21  2010-11-15
    - "package" arg can now tell generate_role to use a specific package (rjbs)

0.20  2010-11-02
    - Minor test refactoring to fix blead support [rt.perl.org #78244]

0.19  2010-07-08
    - Documentation improvements

0.18  2010-03-10
    - Improve the error message when you leave off the role {} block
      http://stackoverflow.com/questions/2418177/moose-and-error-messages-the-sun-and-the-moon/2418429

0.17  2010-02-11
    - MANIFEST fixes (Karen Etheridge)

0.16  2010-02-04
    - Allow specifying custom metaclasses for parameterized
      roles (Oliver Charles)

    - Documentation improvements (Oliver Charles, Sartak)

0.15  2010-01-05
    - Move the guts of MXRPMR::Parameterized into a trait. See
      http://www.nntp.perl.org/group/perl.moose/2010/01/msg1294.html (Sartak)

0.14  2009-12-06
    - Fix a nesting bug reported by nothingmuch with a test case provided
      by rafl (Sartak)

    - Parameterized roles now keep track of their genitor role (Sartak)

    - Delegate parameterizable->has_parameter to parameters->has_attribute
      (Sartak)

0.13  2009-09-11
    - Documentation improvements (Sartak)

0.12  2009-08-15
    - Add MooseX::Role::Parameterized->current_metaclass for forward compat
      (Sartak)

    - Various documentation improvements (Sartak)

0.11  2009-08-09
    - Explicitly test-require Test::Moose for Fedora packaging (RSRCHBOY)

    - Finally a sensible synopsis (Sartak)

    - More examples are always good :) (Sartak)

0.10  2009-06-25
    - Fix test that now warns in Moose (Sartak)

0.09  2009-06-14
    - Give all modules a version number (Sartak)

0.08  2009-06-14
    - Fix mis-packaging :) (Sartak)

0.07  2009-06-14
    - Leave the jules nest for github (Sartak)

    - Switch to Module::Install (Sartak)

0.06  2009-05-12
    - Fix long-standing role-role combination bug, reported by several
      users (Sartak)

    - Documentation improvements (Sartak)

0.05  2009-04-24
    - Use replace_constructor in Meta::Parameter's make_immutable (nothingmuch)


lib/MooseX/Role/Parameterized/Meta/Trait/Parameterizable.pm
package MooseX::Role::Parameterized::Meta::Trait::Parameterizable;
# ABSTRACT: trait for parameterizable roles

our $VERSION = '1.12';

use Moose::Role;
use MooseX::Role::Parameterized::Meta::Role::Parameterized;
use MooseX::Role::Parameterized::Parameters;
use Module::Runtime 'use_module';
use namespace::autoclean;

has parameterized_role_metaclass => (
    is      => 'ro',
    isa     => 'ClassName',
    default => 'MooseX::Role::Parameterized::Meta::Role::Parameterized',
);

has parameters_class => (
    is      => 'ro',
    isa     => 'ClassName',
    default => 'MooseX::Role::Parameterized::Parameters',
);

has parameters_metaclass => (
    is      => 'rw',
    isa     => 'Moose::Meta::Class',
    lazy    => 1,
    builder => '_build_parameters_metaclass',
    handles => {
        has_parameter        => 'has_attribute',
        add_parameter        => 'add_attribute',
        construct_parameters => 'new_object',
    },
    predicate => '_has_parameters_metaclass',
);

has role_generator => (
    is        => 'rw',
    isa       => 'CodeRef',
    predicate => 'has_role_generator',
);

sub _build_parameters_metaclass {
    my $self = shift;

    return $self->parameters_class->meta->create_anon_class(
        superclasses => [$self->parameters_class],
    );
}

my $package_counter = 0;
sub generate_role {
    my $self     = shift;
    my %args     = @_;

    my $parameters = blessed($args{parameters})
                   ? $args{parameters}
                   : $self->construct_parameters(%{ $args{parameters} });

    confess "A role generator is required to apply parameterized roles (did you forget the 'role { ... }' block in your parameterized role '".$self->name."'?)"
        unless $self->has_role_generator;

    my $parameterized_role_metaclass = $self->parameterized_role_metaclass;
    use_module($parameterized_role_metaclass);

    my $package = $args{package};
    unless ($package) {
        $package_counter++;
        $package = $self->name . '::__ANON__::SERIAL::' . $package_counter;
    }
    my $role = $parameterized_role_metaclass->create(
        $package,
        genitor    => $self,
        parameters => $parameters,
    );

    local $MooseX::Role::Parameterized::CURRENT_METACLASS = $role;

    # The generate_role method is being called directly by things like
    # MooseX::ClassCompositor. We don't want to force such modules to pass
    # this arg so we default to something sane.
    my $orig_apply = $args{orig_apply} || Moose::Meta::Role->can('apply');
    $self->$orig_apply($role);

    $self->role_generator->($parameters,
        operating_on => $role,
        consumer     => $args{consumer},
    );

    # don't just return $role here, because it might have been changed when
    # metaroles are applied
    return $MooseX::Role::Parameterized::CURRENT_METACLASS;
}

sub _role_for_combination {
    my $self = shift;
    my $parameters = shift;

    return $self->generate_role(
        parameters => $parameters,
    );
}

around apply => sub {
    my $orig     = shift;
    my $self     = shift;
    my $consumer = shift;
    my %args     = @_;

    my $role = $self->generate_role(
        consumer   => $consumer,
        parameters => \%args,
        orig_apply => $orig,
    );

    $role->apply($consumer, %args);
};

around reinitialize => sub {
    my $orig  = shift;
    my $class = shift;
    my ($pkg) = @_;
    my $meta  = blessed($pkg) ? $pkg : find_meta($pkg);

    my $meta_meta = $meta->meta;

    my %p;
    if ( $meta_meta->can('does_role') && $meta_meta->does_role(__PACKAGE__) ) {
        %p = map { $_ => $meta->$_ }
            qw( parameterized_role_metaclass parameters_class );
        $p{parameters_metaclass} = $meta->parameters_metaclass
            if $meta->_has_parameters_metaclass;
        $p{role_generator} = $meta->role_generator
            if $meta->has_role_generator;
    }

    my $new = $class->$orig(
        @_,
        %p,
    );

    return $new;
};

1;

__END__

=head1 DESCRIPTION

This is the trait that is applied to the metaclass for parameterizable roles,
roles that have their parameters currently unbound. These are the roles that
you use L<Moose/with>, but instead of composing the parameterizable role, we
construct a new parameterized role
(L<MooseX::Role::Parameterized::Meta::Role::Parameterized>) and use that new
parameterized role instead.

=head1 ATTRIBUTES

=head2 parameterized_role_metaclass

The name of the class that will be used to construct the parameterized role.

=head2 parameters_class

The name of the class that will be used to construct the parameters object.

=head2 parameters_metaclass

A metaclass representing this role's parameters. It will be an anonymous
subclass of L</parameters_class>. Each call to
L<MooseX::Role::Parameters/parameter> adds an attribute to this metaclass.

When this role is consumed, the parameters object will be instantiated using
this metaclass.

=head2 role_generator

A code reference that is used to generate a role based on the parameters
provided by the consumer. The user usually specifies it using the
L<MooseX::Role::Parameterized/role> keyword.

=head1 METHODS

=head2 add_parameter $name, %options

Delegates to L<Moose::Meta::Class/add_attribute> on the
L</parameters_metaclass> object.

=head2 construct_parameters %arguments

Creates a new L<MooseX::Role::Parameterized::Parameters> object using metaclass
L</parameters_metaclass>.

The arguments are those specified by the consumer as parameter values.

=head2 generate_role %arguments

This method generates and returns a new instance of
L</parameterized_role_metaclass>. It can take any combination of
three named arguments:

=over 4

=item parameters

A hashref of parameters for the role, same as would be passed in at a "with"
statement.

=item package

A package name that, if present, we will use for the generated role; if not,
we generate an anonymous role.

=item consumer

=for stopwords metaobject

A consumer metaobject, if available.

=back

=head2 apply

Overrides L<Moose::Meta::Role/apply> to automatically generate the
parameterized role.

=cut


lib/MooseX/Role/Parameterized.pm
package MooseX::Role::Parameterized;
# ABSTRACT: Moose roles with composition parameters
# KEYWORDS: moose extension parameter role arguments dynamic parameterised parameterizable parameterisable

our $VERSION = '1.12';

use 5.008001;
use Moose 2.0300 ();
use Moose::Exporter;
use Carp 'confess';
use Moose::Util 'find_meta';
use namespace::clean 0.19;

use MooseX::Role::Parameterized::Meta::Trait::Parameterizable;

our $CURRENT_METACLASS;

sub current_metaclass { $CURRENT_METACLASS }

my $meta_lookup = sub {
    my $for = shift;
    current_metaclass() || find_meta($for);
};

Moose::Exporter->setup_import_methods(
    also           => 'Moose::Role',
    with_caller    => [ 'parameter', 'role' ],
    with_meta      => [ 'method', 'with' ],
    meta_lookup    => $meta_lookup,
    role_metaroles => {
        role => ['MooseX::Role::Parameterized::Meta::Trait::Parameterizable'],
    },
);

sub parameter {
    my $caller = shift;

    confess "'parameter' may not be used inside of the role block"
        if current_metaclass && current_metaclass->genitor->name eq $caller;

    my $meta = find_meta($caller);

    my $names = shift;
    $names = [$names] if !ref($names);

    for my $name (@$names) {
        $meta->add_parameter($name => (
            is => 'ro',
            @_,
        ));
    }
}

sub role (&) {
    my $caller         = shift;
    my $role_generator = shift;

    confess "'role' may not be used inside of the role block"
        if current_metaclass && current_metaclass->genitor->name eq $caller;

    find_meta($caller)->role_generator($role_generator);
}

sub method {
    my $meta = shift;
    my $name = shift;
    my $body = shift;

    my $method = $meta->method_metaclass->wrap(
        package_name => $meta->name,
        name         => $name,
        body         => $body,
    );

    $meta->add_method($name => $method);
}

sub with {
    local $CURRENT_METACLASS = undef;
    Moose::Role::with(@_);
}


1;

__END__

=for Pod::Coverage current_metaclass method parameter role with

=head1 SYNOPSIS

    package Counter;
    use MooseX::Role::Parameterized;

    parameter name => (
        isa      => 'Str',
        required => 1,
    );

    role {
        my $p = shift;

        my $name = $p->name;

        has $name => (
            is      => 'rw',
            isa     => 'Int',
            default => 0,
        );

        method "increment_$name" => sub {
            my $self = shift;
            $self->$name($self->$name + 1);
        };

        method "reset_$name" => sub {
            my $self = shift;
            $self->$name(0);
        };
    };

    package MyGame::Weapon;
    use Moose;

    with Counter => { name => 'enchantment' };

    package MyGame::Wand;
    use Moose;

    with Counter => { name => 'zapped' };

=head1 L<MooseX::Role::Parameterized::Tutorial>

B<Stop!> If you're new here, please read
L<MooseX::Role::Parameterized::Tutorial> for a much gentler introduction.

=head1 DESCRIPTION

Your parameterized role consists of two new things: parameter declarations
and a C<role> block.

Parameters are declared using the
L<MooseX::Role::Parameterized::Tutorial/parameter> keyword which very much
resembles L<Moose/has>. You can use any option that L<Moose/has> accepts. The
default value for the C<is> option is C<ro> as that's a very common case. Use
C<< is => 'bare' >> if you want no accessor. These parameters will get their
values when the consuming class (or role) uses L<Moose/with>. A parameter
object will be constructed with these values, and passed to the C<role> block.

The C<role> block then uses the usual L<Moose::Role> keywords to build up a
role. You can shift off the parameter object to inspect what the consuming
class provided as parameters. You use the parameters to customize your
role however you wish.

There are many possible implementations for parameterized roles (hopefully with
a consistent enough API); I believe this to be the easiest and most flexible
design. Coincidentally, Pugs originally had an eerily similar design.

See L<MooseX::Role::Parameterized::Extending> for some tips on how to extend
this module.

=head2 Why a parameters object?

I've been asked several times "Why use a parameter I<object> and not just a
parameter I<hashref>? That would eliminate the need to explicitly declare your
parameters."

The benefits of using an object are similar to the benefits of using Moose. You
get an easy way to specify lazy defaults, type constraint, delegation, and so
on. You get to use MooseX modules.

=for stopwords metaobject

You also get the usual introspective and intercessory abilities that come
standard with the metaobject protocol. Ambitious users should be able to add
traits to the parameters metaclass to further customize behavior. Please let
me know if you're doing anything viciously complicated with this extension. :)

=head1 CAVEATS

You must use this syntax to declare methods in the role block:
C<< method NAME => sub { ... }; >>. This is due to a limitation in Perl. In
return though you can use parameters I<in your methods>!

=head1 SEE ALSO

L<http://sartak.org/2009/01/parametric-roles-in-perl-5.html>

L<http://sartak.org/2009/05/the-design-of-parameterized-roles.html>

L<http://stevan-little.blogspot.com/2009/07/thoughts-on-parameterized-roles.html>

L<http://perldition.org/articles/Parameterized%20Roles%20with%20MooseX::Declare.pod>

L<http://www.modernperlbooks.com/mt/2011/01/the-parametric-role-of-my-mvc-plugin-system.html>

L<http://jjnapiorkowski.typepad.com/modern-perl/2010/08/parameterized-roles-and-method-traits-redo.html>

L<http://sartak.org/talks/yapc-asia-2009/(parameterized)-roles/>

=for stopwords Joose

L<https://github.com/SamuraiJack/JooseX-Role-Parameterized> - this extension ported to JavaScript's Joose

=cut
